pkg/framework/plugins/nodeutilization/nodeutilization.go

```diff
@@ -523,3 +425,75 @@ func classifyPods(pods []*v1.Pod, filter func(pod *v1.Pod) bool) ([]*v1.Pod, []*
 
 	return nonRemovablePods, removablePods
 }
+
+// AssessNodesUsagesAndThresholds converts the raw usage data into percentage.
+// In case we need to calculate deviations from the average this function does
+// so as well. Thresholds are provided in percentage but they slightly differ
+// in behavior: if no deviation is enabled they are interpreted as is but with
+// deviations enabled they are considered deviations from the node's average.
+// Returns the usage (pct) and the thresholds (pct) for each node.
+func AssessNodesUsagesAndThresholds(
+	rawUsages, rawCapacities map[string]api.ReferencedResourceList,
+	lowSpan, highSpan api.ResourceThresholds,
+	calculateDeviation bool,
+) (map[string]api.ResourceThresholds, map[string][]api.ResourceThresholds) {
+	// first we normalize the node usage from the raw data (Mi, Gi, etc)
+	// into api.Percentage values.
+	usage := normalizer.Normalize(
+		rawUsages, rawCapacities, ResourceUsageNormalizer,
+	)
+
+	// if we are not taking the average and applying deviations to it we
+	// can simply replicate the same threshold across all nodes and return.
+	if !calculateDeviation {
+		thresholds := normalizer.Replicate(
+			slices.Collect(maps.Keys(rawUsages)),
+			[]api.ResourceThresholds{lowSpan, highSpan},
+		)
+		return usage, thresholds
+	}
+
+	// calculate the average usage and then deviate it according to the
+	// user provided thresholds. We also ensure that the value after the
+	// deviation is at least 1%. this call also replicates the thresholds
+	// across all nodes.
+	thresholds := normalizer.Replicate(
+		slices.Collect(maps.Keys(rawUsages)),
+		normalizer.Min(
+			normalizer.Deviate(
+				normalizer.Average(usage),
+				[]api.ResourceThresholds{
+					lowSpan.Negative(),
+					highSpan,
+				},
+			),
+			1,
+		),
+	)
+
+	return usage, thresholds
+}
+
+// ResourceUsageNormalizer is an implementation of a Normalizer that converts a
+// set of resource usages and totals into percentage. This function operates on
+// Quantity Value() for all the resources except CPU, where it uses
+// MilliValue().
+func ResourceUsageNormalizer(usages, totals api.ReferencedResourceList) api.ResourceThresholds {
+	result := api.ResourceThresholds{}
+	for rname, value := range usages {
+		total, ok := totals[rname]
+		if !ok || value == nil || total == nil {
+			continue
+		}
+
+		used, avail := value.Value(), total.Value()
+		if rname == v1.ResourceCPU {
+			used, avail = value.MilliValue(), total.MilliValue()
+		}
+
+		// ensures that the percentage is between 0 and 100.
```

Comment:
This should always be the case. If `used < 0` or `used > avail` the data are invalid. This needs to be handled during data validation. Otherwise, the normalizer might hide it and later on the code might produce incorrect results.

Reply:
Done.